activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -15,47 +15,37 @@ def source_reflection
         def associated_records_by_owner(preloader)
           through_scope = through_scope()
 
-          preloader.preload(owners,
+          through_preloaders = preloader.preload(owners,
                             through_reflection.name,
-                            through_scope)
+                            through_scope,
+                            should_skip_setting_target?(owners, through_reflection.name, through_scope))
 
-          through_records = owners.map do |owner|
-            center = owner.association(through_reflection.name).target
-            [owner, Array(center)]
-          end
-
-          reset_association(owners, through_reflection.name, through_scope)
```

Comment:
Take `A -> B -> C -> P` for example, suppose we are calling `A.preload(:C, :P)`. When preloading `C`, the `A -> C` and `B -> C` will be cached. Then when preloading `P`, the `reset_association` will reset `A -> C` and `B -> C`, wastes the work of preloading `C`. So I prefer to skip over to reset here.